File: aop/aop-annotation-processor/src/test/java/ru/tinkoff/kora/aop/annotation/processor/aoptarget/AopTarget1.java
package ru.tinkoff.kora.aop.annotation.processor.aoptarget;

import com.squareup.javapoet.CodeBlock;
import ru.tinkoff.kora.aop.annotation.processor.KoraAspect;
import ru.tinkoff.kora.aop.annotation.processor.KoraAspectFactory;
import ru.tinkoff.kora.common.AopAnnotation;
import ru.tinkoff.kora.common.Tag;

import javax.annotation.Nullable;
import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.type.TypeKind;
import java.util.List;
import java.util.Optional;
import java.util.Set;

import static com.squareup.javapoet.CodeBlock.joining;

public class AopTarget1 {
    public AopTarget1(String argument, @Tag(String.class) Integer tagged) {}

    public interface ProxyListener1 {
        void call(String annotationValue);
    }

    @AopAnnotation
    public @interface TestAnnotation1 {
        String value();
    }

    public void shouldNotBeProxied1() {}

    void shouldNotBeProxied2() {}

    @TestAnnotation1("testMethod1")
    public String testMethod1() {
        return "test";
    }

    @TestAnnotation1("testMethod2")
    protected void testMethod2(@Nullable String param) {}

    public static class Aspect1Factory implements KoraAspectFactory {

        @Override
        public Optional<KoraAspect> create(ProcessingEnvironment processingEnvironment) {
            return Optional.of(new Aspect1(processingEnvironment));
        }
    }

    public static class Aspect1 implements KoraAspect {
        private final ProcessingEnvironment env;

        public Aspect1(ProcessingEnvironment processingEnvironment) {
            this.env = processingEnvironment;
        }

        @Override
        public Set<String> getSupportedAnnotationTypes() {
            return Set.of(TestAnnotation1.class.getCanonicalName());
        }

        @Override
        public ApplyResult apply(ExecutableElement executableElement, String superCall, AspectContext aspectContext) {
            var annotation = executableElement.getAnnotation(TestAnnotation1.class);
            var field = aspectContext.fieldFactory().constructorParam(
                this.env.getElementUtils().getTypeElement(ProxyListener1.class.getCanonicalName()).asType(),
                List.of()
            );
            var b = CodeBlock.builder()
                .add("this.$L.call($S);\n", field, annotation.value());
            if (executableElement.getReturnType().getKind() != TypeKind.VOID) {
                b.add("return ");
            }

            b.add(executableElement.getParameters().stream()
                .map(p -> CodeBlock.of("$L", p))
                .collect(joining(", ", superCall + "(", ");\n")));

            return new ApplyResult.MethodBody(b.build());
        }
    }
}

File: aop/aop-symbol-processor/src/test/kotlin/ru/tinkoff/kora/aop/ksp/aoptarget/AopTarget1.kt
package ru.tinkoff.kora.aop.ksp.aoptarget

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.getAnnotationsByType
import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.CodeBlock
import ru.tinkoff.kora.aop.symbol.processor.KoraAspect
import ru.tinkoff.kora.aop.symbol.processor.KoraAspectFactory
import ru.tinkoff.kora.common.AopAnnotation
import ru.tinkoff.kora.common.Tag

open class AopTarget1(val argument: String?, @Tag(String::class) val tagged: Int) {
    interface ProxyListener1 {
        fun call(annotationValue: String?)
    }

    @AopAnnotation
    annotation class TestAnnotation1(val value: String)

    fun shouldNotBeProxied1() {}

    fun shouldNotBeProxied2() {}

    @TestAnnotation1("testMethod1")
    open fun testMethod1(): String? {
        return "test"
    }

    @TestAnnotation1("testMethod2")
    open fun testMethod2(param: String?) {
    }

    @KspExperimental
    class Aspect1Factory : KoraAspectFactory {
        override fun create(resolver: Resolver): KoraAspect {
            return Aspect1(resolver)
        }
    }

    @KspExperimental
    class Aspect1(private val resolver: Resolver) : KoraAspect {

        override fun getSupportedAnnotationTypes(): Set<String> {
            return setOf(TestAnnotation1::class.java.canonicalName)
        }

        override fun apply(ksFunction: KSFunctionDeclaration, superCall: String, aspectContext: KoraAspect.AspectContext): KoraAspect.ApplyResult {
            val annotation = ksFunction.getAnnotationsByType(TestAnnotation1::class).first()
            val field: String = aspectContext.fieldFactory.constructorParam(
                resolver.getClassDeclarationByName(ProxyListener1::class.java.canonicalName)!!.asType(listOf()),
                listOf()
            )
            val b = CodeBlock.builder()
                .add("this.%L.call(%S)\n", field, annotation.value)
            if (ksFunction.returnType != resolver.builtIns.unitType) {
                b.add("return ")
            }
            b.add(ksFunction.parameters
                .map { p -> CodeBlock.of("%L", p) }
                .joinToString(", ", "$superCall(", ")\n")
            )
            return KoraAspect.ApplyResult.MethodBody(b.build())
        }
    }
}
